shaders/src/shading_lit.fs

```diff
@@ -2,30 +2,44 @@
 // Lighting
 //------------------------------------------------------------------------------
 
-float computeDiffuseAlpha(float a) {
-#if defined(BLEND_MODE_TRANSPARENT) || defined(BLEND_MODE_FADE) || defined(BLEND_MODE_MASKED)
-    return a;
-#else
-    return 1.0;
-#endif
-}
-
 #if defined(BLEND_MODE_MASKED)
 float computeMaskedAlpha(float a) {
     // Use derivatives to smooth alpha tested edges
     return (a - getMaskThreshold()) / max(fwidth(a), 1e-3) + 0.5;
 }
-#endif
+
+float computeDiffuseAlpha(float a) {
+    // If we reach this point in the code, we already know that the fragment is not discarded due
+    // to the threshold factor. Therefore we can just output 1.0, which prevents a "punch through"
+    // effect from occuring. We do this only for TRANSLUCENT views in order to prevent breakage
+    // of ALPHA_TO_COVERAGE.
+    if (frameUniforms.needsAlphaChannel == 1.0) {
+        return 1.0;
+    }
+    return a;
```

Comment:
you could actually do `return frameUniforms.needsAlphaChannel*(1 - a) + a;` to avoid the conditional. Not sure if it would be best :-)